### ISSUE TITLE
Unknown Column Path

### ISSUE THREAD

**Issue by AlexanderMann** [datamill-co/target-postgres#99]
# Motivation

While using `tap-hubspot`, Singer Slack User Patryk Kallinowski ran into the following error:

```
target_postgres.postgres.PostgresError: ('Exception writing records', Exception("Unknown column path: ('properties', 'num_contacted_notes', 'value') for table: ('deals',)"))
```

After uploading his `catalog.json` file generated by `tap-hubspot`, the issue appears to be around this stream/sub-schema:

```json
{
  "streams": [
    {
      "stream": "deals",
      "tap_stream_id": "deals",
      "schema": {
        "type": "object",
        "properties": {
          ...
          "properties": {
            "type": "object",
            "properties": {
              ...
              "num_contacted_notes": {
                "type": "object",
                "properties": {
                  "value": {
                    "type": ["null", "number", "string"]
                  },
                  ...
                }
              },
              ...
            }
          }
        }
      },
      "metadata": [
        ...
      ]
    }
  ]
}
```

At the very least, this error message can be improved to:

- contain more context and information for those debugging
- more descriptive/better language to identify what precisely has gone wrong
- be associated with a function which has a complete docstring

And:

- more testing to assure that whatever underlying bug (if any) is resolved

## Notes

https://singer-io.slack.com/archives/C2TGFCZEV/p1552316668073400

`catalog.json` can't be uploaded as a part of this issue, available via download in the link above.

## Suggested Musical Pairing

https://www.youtube.com/watch?v=PfrHCNo2I3M

**Comment by AlexanderMann** [comment]
I think I've identified the bug here. It has to do with how we're serializing fields which have _multiple_ types. There are, I think, a couple bugs therein:

- multi type, non numeric
- single type, `[null, number]` which has values being sent as `integer`s (ie, the schema is `[null, number]`, but the values are something like `1234, 567, 8, 90` etc.

I'm fixing this bug, and _then_ I will improve the error message.

**Comment by AlexanderMann** [comment]
Correction: the single type for `[null, number]` is perfectly fine.

I've found the issue. When we have a `null` value we _cannot_ assign it a "python type". ie, we don't _really_ know what value "lack of value" represents in Postgres. As such, for split column types, we don't know which column to write "nothing" to.

The fix, is effectively to do _less_ work. Presently we set default `null` values for _every_ column in the remote schema. Since we've already done this, if we see that we're going to write yet _another_ `null` value to the column, we can avoid this, which circumvents us trying to find a column mapping for an unknown schema.
